src/inspector/main_thread_interface.h

```diff
@@ -0,0 +1,137 @@
+#ifndef SRC_INSPECTOR_MAIN_THREAD_INTERFACE_H_
+#define SRC_INSPECTOR_MAIN_THREAD_INTERFACE_H_
+
+#if !HAVE_INSPECTOR
+#error("This header can only be used when inspector is enabled")
+#endif
+
+#include "env.h"
+#include "inspector_agent.h"
+#include "node_mutex.h"
+
+#include <deque>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace v8_inspector {
+class StringBuffer;
+class StringView;
+}  // namespace v8_inspector
+
+namespace node {
+namespace inspector {
+class Request {
+ public:
+  virtual void Call() = 0;
+  virtual ~Request() {}
+};
+
+template <typename Target, typename Arg>
+class SingleArgumentFunctionCall : public Request {
+ public:
+  using Fn = void (Target::element_type::*)(Arg);
+
+  SingleArgumentFunctionCall(Target target, Fn fn, Arg argument)
+                             : target_(target),
+                               fn_(fn),
+                               arg_(std::move(argument)) {}
+
+  void Call() override {
+    Apply(target_, fn_, std::move(arg_));
+  }
+
+ private:
+  template <typename Element>
+  void Apply(std::shared_ptr<Element> target, Fn fn, Arg arg) {
+    if (target != nullptr)
+      (target.get()->*fn)(std::move(arg));
+  }
+
+  template <typename Element>
+  void Apply(std::weak_ptr<Element> target, Fn fn, Arg arg) {
+    Apply(target.lock(), fn, arg);
+  }
+
+  Target target_;
+  Fn fn_;
+  Arg arg_;
+};
```

Comment:
It makes calling the `Post` overload beload a bit awkward, but it would be nice if we could use something less complex and more generic than this here. Using a `std::function<void()>` as the base type works if we accept having to manually transfer `unique_ptr` contents:

<details><summary> diff in the fold </summary>

```diff
diff --git a/src/inspector/main_thread_interface.cc b/src/inspector/main_thread_interface.cc
index 66ec77fe3bbb..390f5f8445cd 100644
--- a/src/inspector/main_thread_interface.cc
+++ b/src/inspector/main_thread_interface.cc
@@ -109,15 +109,16 @@ class CrossThreadInspectorSession : public InspectorSession {
       : handle_(handle),
         block_(std::make_shared<CrossThreadInspectorSessionBlock>(
             handle_, prevent_shutdown)) {
-    handle_->Post(block_,
-                  &CrossThreadInspectorSessionBlock::Connect,
-                  std::move(delegate));
+    handle_->Post([block = block_, delegate = delegate.release()]() {
+      block->Connect(std::unique_ptr<InspectorSessionDelegate>(delegate));
+    });
   }
 
   void Dispatch(StringView message) override {
-    handle_->Post(block_,
-                  &CrossThreadInspectorSessionBlock::Dispatch,
-                  StringBuffer::create(message));
+    handle_->Post([block = block_,
+                   message = StringBuffer::create(message).release()]() {
+      block->Dispatch(std::unique_ptr<StringBuffer>(message));
+    });
   }
 
  private:
diff --git a/src/inspector/main_thread_interface.h b/src/inspector/main_thread_interface.h
index dd815ca8db0f..e68b9612da4b 100644
--- a/src/inspector/main_thread_interface.h
+++ b/src/inspector/main_thread_interface.h
@@ -27,35 +27,16 @@ class Request {
   virtual ~Request() {}
 };
 
-template <typename Target, typename Arg>
-class SingleArgumentFunctionCall : public Request {
+class FunctionCall : public Request {
  public:
-  using Fn = void (Target::element_type::*)(Arg);
-
-  SingleArgumentFunctionCall(Target target, Fn fn, Arg argument)
-                             : target_(target),
-                               fn_(fn),
-                               arg_(std::move(argument)) {}
+  explicit FunctionCall(std::function<void()> fn) : fn_(fn) {}
 
   void Call() override {
-    Apply(target_, fn_, std::move(arg_));
+    fn_();
   }
 
  private:
-  template <typename Element>
-  void Apply(std::shared_ptr<Element> target, Fn fn, Arg arg) {
-    if (target != nullptr)
-      (target.get()->*fn)(std::move(arg));
-  }
-
-  template <typename Element>
-  void Apply(std::weak_ptr<Element> target, Fn fn, Arg arg) {
-    Apply(target.lock(), fn, arg);
-  }
-
-  Target target_;
-  Fn fn_;
-  Arg arg_;
+  std::function<void()> fn_;
 };
 
 std::unique_ptr<v8_inspector::StringBuffer> Utf8ToStringView(
@@ -75,10 +56,8 @@ class MainThreadHandle : public std::enable_shared_from_this<MainThreadHandle> {
       std::unique_ptr<InspectorSessionDelegate> delegate,
       bool prevent_shutdown);
   void Post(std::unique_ptr<Request> request);
-  template <typename Target, typename Fn, typename Arg>
-  void Post(Target target, Fn fn, Arg argument) {
-    Post(std::unique_ptr<Request>(new SingleArgumentFunctionCall<Target, Arg>(
-        target, fn, std::move(argument))));
+  void Post(std::function<void()> function) {
+    Post(std::unique_ptr<Request>(new FunctionCall(function)));
   }
 
 
```

</details>

Reply:
I do not believe this change would make the code more readable, especially considering tricks for the unique_ptr support. There is at least one more usage of this method in the Workers debug PR so I feel like this overload works ok...

Reply:
Yeah, no problem – I can always open a separate PR if I manage to come up with something better.